prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+                    "function": "logAtag",
+                    "args": {
+                      "analyticsValue": "control"
+                    }
+                  }
+                ]
+
+              }
+            ]
+          }
+        ]
+      }
+    }
+  }
+}
+```
+
+## How it Works
+
+Similar to the Floors feature, it's assumed that an offline machine-learning algorithm will be part of a bidder optimization feedback loop to adjust to changing conditions. Here's the assumed architecture:
+
+![Prebid Server Rules Engine Module](/assets/images/prebid-server/pbs_rules_engine_context.png){:class="pb-xlg-img"}
+
+1. Auction requests come in that trigger the Rules Engine Module.
+1. The module reads the current set of rules from account configuration or by directly fetching them from the service provider.
+1. The results of the optimization decision and the auction are logged via analytics.
+1. Eventually the configuration is updated with refined rules.
+
+The initial goal of this module is to define the runtime capabilities that are required to surgically select bidders. We've defined a technical interface and a runtime behavior. Some uses of the module might be manually configured, but the most valuable use of the system will be in the periodic offline machine learning. The goal there is for Prebid to step out of the way and let members of the ecosystem compete to do the best job.
+
+We envision a time when, like the floors sytem, there are multiple vendors providing optimization services of different kinds.
+
+## Configuration Details
+
+Configuration is covered in several sections:
+
+1. [Module configuration syntax](#module-account-level-configuration) - the JSON used to express which bidders should run when. This is basically a kind of a Domain-Specific Language (DSL) where the 
+1. [Rule Trees](#rule-trees) - groups of rules should be thought of as 'trees' split along dimensions like 'datacenter' and 'channel'. These trees are processed in a particular way.
+1. [Schema Functions](#schema-functions) - which dimensions and comparisons can be used in the configuration. 
+1. [Results Functions](#results-functions) - the specific outcomes once a specific rule is chosen.
+1. [Runtime Processing](#runtime-processing) - how it all comes together.
+
+### Module Account-Level Configuration
+
+These are the full set of attributes that can be present in the account-level configuration:
+
+{: .table .table-bordered .table-striped }
+| Property | Scope | Type | Description | Example |
+|----------|-------|------|-------------|---------|
+| enabled | optional | boolean | Whether the module as a whole is enabled. Allows for easy temporary deactivation. Defaults to true. | true |
+| generateRulesFromBidderConfig | optional | boolean | (PBS-Go only) If true, the module automatically generates rules based on 'geoscope' section of the bidder YAML files to include or exclude that bidder from certain countries. Defaults to true. | false |
```

Comment:
I believe it makes sense to default this true as you have mentioned here right? If you're using the rules engine module why wouldn't you want to turn on the geoscoping rules?

Reply:
Up to you guys. I don't think it's supported in PBS-J.